### ISSUE TITLE
Invalid gpx file

### ISSUE THREAD

**Issue by grinsekatze003** [iam4x/pokemongo-webspoof#125]
Pokemongo Webspoof works fine but after approx. 10 minutes I'm getting this error in xcode:

> Location simulation cannot continue because the GPX file "pokemonLocation.gpx" is invalid. Failed to parse GPX file (/var/folders/qz/s3n_2lqj0fs_hgqt7wyzy0400000gn/T/tmp-50059l4kLImVu2lAc/pokemonLocation.gpx)

What is wrong?

**Comment by AbelIngrand** [comment]
This happens when Xcode reads the gpx file at the same time that the app is writing it. Just reseller pokemonLocation from the debug menu and it should work again

**Comment by grinsekatze003** [comment]
@AbelIngrand The problem is when this happens, the app jumps back to my real location and this could lead to bans. I think we should find a solution for this?

**Comment by AbelIngrand** [comment]
Looks like a race condition which is tough to get over with the current app logic. A solution could be to not spam the arrow keys. As I see it, this happens if you update the location two times in a short amount of time:
The first update triggers the script that tells Xcode to simulate the new location but before it can read it, it as been opened by the second update that writes the new location at the same moment.

The solution I see is to use a Lock like they do in Pokemon Go keyboard that prevents writing to the gpx while Xcode is reading it.

@iam4x do you think you could look into this ?

**Comment by d3viant0ne** [comment]
@grinsekatze003 - @AbelIngrand is correct, it's a race condition. His solution is also correct. This race condition is also where the app is leaking memory which will eventually cause it to crash.

**Comment by iam4x** [comment]
This is a weird behaviour, to handle the race condition I throttled the write of `pokemonLocation.gpx` to one seconds. You can see code here: https://github.com/iam4x/pokemongo-webspoof/blob/master/src/models/user-location.js#L39-L74

Seems like one second is not enough for slow computer? Which is weird...

**Comment by grinsekatze003** [comment]
Hmm, I have the latest MacBook Air, so it shouldn't be a problem that my computer is too slow?

**Comment by Godrek** [comment]
I get this error immediately whenever I try to simulate location. 

Also have a relatively new macbook pro so speed is definitely not an issue.

**Comment by iam4x** [comment]
I suspect that the coordinates I generate are wrong sometimes but it's only kind of warning should work anyway after next move.

@Godrek where are you located when you experience this?

**Comment by Godrek** [comment]
@iam4x I'm in Bellevue, Washington. 

Repro steps for me:
Start Pokemon Go
Open xcode project
Build + deploy onto phone
Try to set location to "pokemonLocation" and then the error appears

I also don't get a client on the macbook I'm building from, isn't a window supposed to pop up that lets me move, etc.?

**Comment by evannudd** [comment]
To help prevent accidental soft ban I recommend you don't start Pokemon go until the end of this process.

You should not be opening the xCode project. If you downloaded the .app.tar.gz then unzip that and run the .app inside that. This should open the window you're looking for, as well as will open the xCode workspace.

Then follow the instructions on getting Xcode location simulation enabled and make sure the marker is where you want it on the app. Then open Pokemon Go, and you should appear in the spoofed location.

**Comment by SaipanDamashii** [comment]
I get the invalid gpx file at the very beginning, when I try to select the location simulation to se it to PokemonLocation.

**Comment by AbelIngrand** [comment]
Yes, this is normal, try moving in the map a little then select pokemonLication. You can then activate auto update in Xcode

**Comment by d3viant0ne** [comment]
@SaipanDamashii - When you launch, the gpx file doesn't have a location set. Selecting "Auto Update Xcode Project" and moving a few key clicks will write a location to the file & it is then valid as @AbelIngrand suggested.

Then turn off Auto Update and follow the normal application startup procedure from start to finish as this is only necessary at initial application launch.

**Comment by SaipanDamashii** [comment]
Ahhhhh. When I launched it, there was no option to click on Auto update Xcode project. Is that in a menu somewhere, or is it supposed to pop up, etc.?

**Comment by ksulli10** [comment]
I am having the same problem... there is no option to Auto Update Xcode Project that I can see.  It says "Fail to parse" on the gpx file.

**Comment by lanabug** [comment]
Same here @ksulli10 
I also do not have an .app to open like @evannudd is referring to.

**Comment by joshpalan** [comment]
@d3viant0ne I'm still having the gpx file is invalid issue, though I've noticed that my changes made in Xcode do not save unless I open the actual project files from inside the app folder. It seems the files are completely rewritten every time I launch the app. Is there a way to save the gpx file so that I don't need to click a few times every run? Does the gpx file need to be blank at launch?

**Comment by jlearman** [comment]
I get the gpx file is invalid as well, but I understand why: I have no idea how to run the "app" that creates the gpx file.  So, the gpx file is blank, and obviously xcode can't open  it.  How do I run the "app"? Sorry... noob here...

**Comment by ksulli10** [comment]
@lanabug @joshpalan @jlearman Make sure you guys are clicking the bullet point download under the GIF in the description, labeled "Download last release pokemongo-webspoof.app.tar.gz (v1.2.0)".  Don't click "Clone and Download" which is what I was doing.  When you download the .gz, click it to untar the file and there should just be a .app file sitting there.  Double-click to open and a GUI will pop up, as well as open Xcode.  The GUI is where you click "Auto update xcode."  Rename the bundle identifier in xcode, build, and run with your USB-connected iPhone.  It should work now without the "invalid GPX" error message.

**Comment by joshpalan** [comment]
@ksulli10 That's what I did, however every time I ran the app I would have to reenter the bundle identifier. It was like the file was being written at runtime from the pokemongo-webspoof.app/Contents/Resources/app/xcode-project folder. So I actually did a show package contents, changed these files, saved them, then ran the app and the bundle identifier was saved. This is what leads me to wonder if the gpx file is supposed to be blank at runtime or not.
